File: 2_ruby/parentesis_balanceados.rb
def parentesis_balanceados?(cadena)
  apertura = "([{"
  cierre = ")]}"
  comparador = {")" => "(", "]" => "[", "}" => "{"}
  pila = []
  cadena.chars.each do |caracter|
    if apertura.include? caracter
      pila.append(caracter)
    elsif cierre.include? caracter
      if pila.length == 0
        return false
      end
      if pila[-1] == comparador[caracter]
        pila.pop
      else
        return false
      end
    end
  end
  return pila.length == 0
end

p parentesis_balanceados?("{[()]}")
p parentesis_balanceados?("{[(]}")

File: 1_python/parentesis_balanceados.py
def parentesis_balanceados(cadena):
    apertura = "([{"
    cierre = ")]}"
    comparador = {")": "(", "]": "[", "}": "{"}
    pila = []
    for caracter in cadena:
        if caracter in apertura:
            pila.append(caracter)
        elif caracter in cierre:
            if len(pila) == 0:
                return False
            if pila[-1] == comparador[caracter]:
                pila.pop()
            else:
                return False
    return len(pila) == 0

print(parentesis_balanceados("{[()]}"))
print(parentesis_balanceados("{[(]}"))

